### PR TITLE
nvJpegDecoderCrop, nvJpegDecoderRandomCrop, nvJpegDecoderSlice

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

### CODE REVIEW

**Comment by JanuszL** [comment]
I think you can extend dali/test/python/test_random_bbcrop.py by adding nvJPEG slice and see if that works.

**Comment by jantonguirao** [comment]
Build 663427

**Comment by jantonguirao** [comment]
Build 670638

**Comment by jantonguirao** [comment]
Build 670651

File: dali/pipeline/pipeline.cc [hunk 1]
```diff
@@ -120,9 +120,23 @@ void Pipeline::AddOperator(OpSpec spec, const std::string& inst_name) {
   //               - make it as an arg of nvJPEGDecoderNew
 
   // nvJGPEGDecoder operator require a special case to be divided in two stages (CPU and Mixed-GPU)
-  const bool nvjpeg_splitting = spec.name() == "nvJPEGDecoderSplitted";
-  if (nvjpeg_splitting) {
-    AddSplittedNvJpegDecoder(spec, inst_name);
+  std::string fullname = spec.name();
```
Comment (jantonguirao) [inline dali/pipeline/pipeline.cc]: note: There is a plan to remove all those "Splitted" operators and use an option in nvJPEGDecoder: https://jirasw.nvidia.com/browse/DALI-603
In that task I'll refactor this code to something more generic

File: dali/benchmark/decoder_bench.cc [hunk 2]
```diff
@@ -308,5 +311,89 @@ BENCHMARK_REGISTER_F(DecoderBench, HostDecoderSlice)->Iterations(100)
...
+      .AddInput("raw_jpegs", "cpu")
+      .AddOutput("images", "gpu"));
+}
+
+BENCHMARK_REGISTER_F(DecoderBench, nvJPEGDecoderRandomCrop)->Iterations(100)
+->Unit(benchmark::kMillisecond)
+->UseRealTime()
+->Apply(PipeArgs);
+
+BENCHMARK_DEFINE_F(DecoderBench, nvJPEGDecoderCrop)(benchmark::State& st) { // NOLINT
```
Comment (mzient) [inline dali/benchmark/decoder_bench.cc]: I think these `// NOLINT`s are the remnant of the dark times when we bowed to cpplint and it's grudge held against non-const l-value references.

Comment (jantonguirao) [reply dali/benchmark/decoder_bench.cc]: Removed (all from this file)

**Comment by jantonguirao** [comment]
Build 670698

File: dali/pipeline/pipeline.cc [hunk 3]
```diff
@@ -120,9 +120,23 @@ void Pipeline::AddOperator(OpSpec spec, const std::string& inst_name) {
...
+  // TODO(janton): make this cleaner and smarter
+  if (fullname == "nvJPEGDecoderSplitted") {
+    AddSplittedNvJpegDecoder(spec, inst_name, fullname, "nvJPEGDecoderCPUStage",
+                             "nvJPEGDecoderGPUStage");
+    return;
+  } else if (fullname == "nvJPEGDecoderSplittedCrop") {
+    AddSplittedNvJpegDecoder(spec, inst_name, fullname, "nvJPEGDecoderCPUStageCrop",
+                             "nvJPEGDecoderGPUStage");
```
Comment (Kh4L) [inline dali/pipeline/pipeline.cc]: `nvJPEGDecoderGPUStageCrop` etc. ?

**Pull request merged**